parquet-variant-compute/src/variant_parser.rs

```diff
@@ -0,0 +1,669 @@
+            | PrimitiveType::Double
+            | PrimitiveType::Decimal8
+            | PrimitiveType::TimestampNtz
+            | PrimitiveType::TimestampLtz => Some(8),
+            PrimitiveType::Decimal16 => Some(16),
+            PrimitiveType::Binary | PrimitiveType::String => None, // Variable length, need to read from data
+        }
+    }
+
+    /// Extract short string data from value bytes
+    pub fn extract_short_string_data(value_bytes: &[u8]) -> Result<&[u8], ArrowError> {
+        if value_bytes.is_empty() {
+            return Err(ArrowError::InvalidArgumentError(
+                "Empty value bytes".to_string(),
+            ));
+        }
+
+        let header = Self::parse_short_string_header(value_bytes[0])?;
+
+        if value_bytes.len() < 1 + header.length {
+            return Err(ArrowError::InvalidArgumentError(format!(
+                "Short string data length {} exceeds available bytes",
+                header.length
+            )));
+        }
+
+        Ok(&value_bytes[1..1 + header.length])
+    }
+
+    /// Extract primitive data from value bytes
+    pub fn extract_primitive_data(value_bytes: &[u8]) -> Result<&[u8], ArrowError> {
+        if value_bytes.is_empty() {
+            return Err(ArrowError::InvalidArgumentError(
+                "Empty value bytes".to_string(),
+            ));
+        }
+
+        let primitive_type = Self::parse_primitive_header(value_bytes[0])?;
+        let data_length = Self::get_primitive_data_length(&primitive_type);
+
+        match data_length {
+            Some(0) => {
+                // Fixed-length 0-byte types (null/true/false)
+                Ok(&[])
+            }
+            Some(len) => {
+                // Fixed-length types with len bytes
+                if value_bytes.len() < 1 + len {
+                    return Err(ArrowError::InvalidArgumentError(format!(
+                        "Fixed length primitive data length {} exceeds available bytes",
+                        len
+                    )));
+                }
+                Ok(&value_bytes[1..1 + len])
+            }
+            None => {
+                // Variable-length types (binary/string) - read length from data
+                if value_bytes.len() < 5 {
+                    return Err(ArrowError::InvalidArgumentError(
+                        "Not enough bytes for variable length primitive".to_string(),
+                    ));
+                }
+                let length = u32::from_le_bytes([
+                    value_bytes[1],
+                    value_bytes[2],
+                    value_bytes[3],
+                    value_bytes[4],
+                ]) as usize;
+                if value_bytes.len() < 5 + length {
+                    return Err(ArrowError::InvalidArgumentError(
+                        "Variable length primitive data exceeds available bytes".to_string(),
+                    ));
+                }
+                Ok(&value_bytes[5..5 + length])
```

Comment:
decoder.rs has utilities could make this code a lot simpler...
```suggestion
                let length_bytes = slice_from_slice_at_offset(value_bytes, 1, 0..4)?;
                let length = unpack_int(length_bytes, 4)?;
                slice_from_slice_at_offset(value_bytes, 5, 0..length)
```